Conjurer Pack

SevenDayCandle opened the pull request:
This pull request contains the 10 cards for the Conjurer.

I also put in a fix for the patch I mentioned yesterday. Feel free to drop it if an alternate solution was already found.

modargo commented:
In addition to the card art, could you add a pack preview image? That's what people will see when your pack shows up in the random selection or the draft, and it's also what gets used in the pack filtering menu (which is a good place to preview how it looks and check that you set up correctly)..

The pack preview image is formatted the same as a skill card and goes in the cards folder with everything else. Like with normal cards, you'll need both the 250x180 and 500x360 versions. You can see a variety of examples like ShamanPack.png/ShamanPack_p.png or WarlockPack.png/WarlockPack_p.png.

modargo commented on src/main/java/thePackmaster/cards/conjurerpack/Enervate.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    public void use(AbstractPlayer p, AbstractMonster m) {
+        if (m != null)
```
Since this isn't in an action, there will be a few oddities -- for example, this will calculate strength for Gremlin Nob before  Nob's Enrage is applied. So if Nob is at 4 Strength with 3 Enrage:
* This sees Nob at 4 Strength and locks in -4 strength
* Nob gains 3 Strength, going up to 7
* Nob loses 4 Strength, going down to 3

It would be more consistent if this worked like so:
* This card is played
* Nob gains 3 Strength, going up to 7
* This applies -5 Strength to Nob, going down to 2

You can achieve this by changing this code to the following:
```
        this.addToBot(new AbstractGameAction() {
            @Override
            public void update() {
                if (m != null)
                {
                    int decAmount = magicNumber;
                    AbstractPower strPower = m.getPower(StrengthPower.POWER_ID);
                    if (strPower != null && strPower.amount > 0)
                    {
                        decAmount += strPower.amount / 2;
                    }
                    this.addToTop(new ApplyPowerAction(m, p, new StrengthPower(m, -decAmount), -decAmount));
                }
                this.isDone = true;
            }
        });
```

modargo commented on src/main/java/thePackmaster/cards/conjurerpack/GeothermalHeating.java:
```diff
@@ -0,0 +1,44 @@
+        blck();
+        if (m != null)
+        {
+            int count = 1;
```
Starting this at 1 means that you'll always get additional block from this, even if the enemy has no debuffs. I think you can start at 0.

You can also remove the `m.powers.size() > 0` check right below this -- it isn't needed, since if `m.powers` is empty, the for loop will happen 0 times.

modargo commented:
Thanks for submitting this pack!

I read through the code and also checked out this PR locally to play around with things. The cards make sense as a whole. In terms of balance, the Petra buff seems like it could be powerful, but there aren't a ton of ways of getting it, so it might turn out to be balanced. Hard to say without playtesting. I also left comments on some issues I noticed and tried to describe what a good way to fix each one would be.

Once you've addressed those issues I'll take a second pass over the PR :smile:.

SevenDayCandle commented on src/main/java/thePackmaster/powers/conjurerpack/PetraPower.java:
```diff
@@ -0,0 +1,56 @@
+
+    @Override
+    public void atEndOfRound() {
+        Wiz.reducePower(this);
```
Thanks for catching that. I intend it to now be reduced by 1 each turn, but I forgot to update the description.

SevenDayCandle commented:
I've addressed the issues you have posted above. Thank you for reviewing!

I'm a little divided on whether Enervate should apply before or after things like Gremlin Nob (i.e. it could be interpreted as the target losing half of its current Strength), but I will make the change for now to see how things pan out, and this would be more beneficial for the player anyhow.

modargo commented:
> I've addressed the issues you have posted above. Thank you for reviewing!
> 
> I'm a little divided on whether Enervate should apply before or after things like Gremlin Nob (i.e. it could be interpreted as the target losing half of its current Strength), but I will make the change for now to see how things pan out, and this would be more beneficial for the player anyhow.

For Enervate, if you do explicitly want it to apply before Nob's Enrage and stuff like that, you may be able to have that happen by doing `this.addToTop(new AbstractGameAction() {...` (instead of `addToBot`). If you end up wanting to make that change it's fine for it to be in a future PR, no need to delay merging this one for it.

SevenDayCandle commented on src/main/java/thePackmaster/actions/conjurerpack/PlayRandomCardAction.java:
```diff
@@ -0,0 +1,51 @@
+            {
+                group.group.remove(c);
+                AbstractDungeon.getCurrRoom().souls.remove(c);
+                this.addToBot(new NewQueueCardAction(c, target, false, true));
```
I had ended up making some additional changes because the previous version had a bug where cards could be chosen twice by the loop. Accordingly, the tmp group was never actually needed, so I removed it.

Finally, I decided to only set the target x/y variables so that the player would see the cards move from the discard pile to slightly left of the center of the screen. Would it be better to just have them start off in the center as was suggested above?

modargo replied:
Setting only target_x/y is fine too. As you've noted, that just means there's the card is moved from the discard to the center of the screen, so it's just a matter of preference for how you want the visual effect to look.
